service/history/handler.go

```diff
@@ -1854,6 +1854,126 @@ func (h *Handler) UpdateWorkflowExecution(
 	return engine.UpdateWorkflowExecution(ctx, request)
 }
 
+func (h *Handler) ReadHistoryBranch(
+	ctx context.Context,
+	request *historyservice.ReadHistoryBranchRequest,
+) (_ *historyservice.ReadHistoryBranchResponse, retError error) {
+	defer log.CapturePanic(h.logger, &retError)
+	h.startWG.Wait()
+
+	if h.isStopped() {
+		return nil, errShuttingDown
+	}
+
+	shardContext, err := h.controller.GetShardByID(request.ShardId)
+	if err != nil {
+		return nil, h.convertError(err)
+	}
+
+	engine, err := shardContext.GetEngine(ctx)
```

Comment:
Maybe there are some reasons I can't see in this repo, but I don't quite follow why those requests need to go through shardContext and historyEngine. Those requests originally can be served even when there's no shard availability.

Reply:
Chatted offline already. The gist of the conclusion is we should wait for the shard to come online before accessing the shard. Even though today technically we can access the persistence data regardless if the shard is available, it is the more correct logic.